File: Sep/Day17/design-a-food-rating-system.cc
#include <bits/stdc++.h>
using namespace std;

class FoodRatings {
    unordered_map<string, pair<int, string>> foodMap; // food -> {rating, cuisine}
    unordered_map<string, priority_queue<pair<int, string>>> cuisineMap; // cuisine -> maxHeap {(-rating, food)}

public:
    FoodRatings(vector<string>& foods, vector<string>& cuisines, vector<int>& ratings) {
        for (int i = 0; i < foods.size(); i++) {
            string food = foods[i], cuisine = cuisines[i];
            int rating = ratings[i];
            foodMap[food] = {rating, cuisine};
            cuisineMap[cuisine].push({-rating, food}); // min-heap trick: negative rating
        }
    }

    void changeRating(string food, int newRating) {
        auto [oldRating, cuisine] = foodMap[food];
        foodMap[food] = {newRating, cuisine};
        cuisineMap[cuisine].push({-newRating, food}); // lazy push
    }

    string highestRated(string cuisine) {
        auto &heap = cuisineMap[cuisine];
        while (!heap.empty()) {
            auto [negRating, food] = heap.top();
            int actualRating = foodMap[food].first;
            if (-negRating == actualRating) return food;
            heap.pop(); // stale entry
        }
        return "";
    }
};

File: Sep/Day17/design-a-food-rating-system.py
class FoodRatings:

    def __init__(self, foods: List[str], cuisines: List[str], ratings: List[int]):
        self.foodMap = {}  # food -> (rating, cuisine)
        self.cuisineMap = defaultdict(list)  # cuisine -> max heap [(-rating, food)]

        for i in range(len(foods)):
            food, cuisine, rating = foods[i], cuisines[i], ratings[i]
            self.foodMap[food] = (rating, cuisine)
            heapq.heappush(self.cuisineMap[cuisine], (-rating, food))  # maxHeap
        

    def changeRating(self, food: str, newRating: int) -> None:
        oldRating, cuisine = self.foodMap[food]
        # update foodMap
        self.foodMap[food] = (newRating, cuisine)
        # push new rating into cuisine heap (lazy update)
        heapq.heappush(self.cuisineMap[cuisine], (-newRating, food))
            
        

    def highestRated(self, cuisine: str) -> str:
        heap = self.cuisineMap[cuisine]
        # Lazy deletion: pop until the top matches the actual rating
        while heap:
            rating, food = heap[0]
            if -rating == self.foodMap[food][0]:  # valid entry
                return food
            heapq.heappop(heap)  # discard stale entry
        


# Your FoodRatings object will be instantiated and called as such:
# obj = FoodRatings(foods, cuisines, ratings)
# obj.changeRating(food,newRating)
# param_2 = obj.highestRated(cuisine)